### PR TITLE
Flow exception details between orchestrator and activity functions

### PR DESCRIPTION
**Pull request by cgillum**
Fixes https://github.com/Azure/azure-functions-durable-extension/issues/84. The new behavior is that activity and sub-orchestrator functions will continue to throw `FunctionFailedException`, but the `InnerException` will now contain the details of the original exception.

This required taking a new version of **DurableTask.Core**, which has fixes for https://github.com/Azure/azure-functions-durable-extension/issues/146 as well.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/MessagePayloadDataConverter.cs [hunk 1]
```diff
@@ -56,6 +53,20 @@ public string Serialize(object value, int maxSizeInKB)
...
+                int payloadSizeInKB = (int)(Encoding.Unicode.GetByteCount(serializedJson) / 1024.0);
+                if (payloadSizeInKB > maxSizeInKB)
+                {
+                    throw new ArgumentException(
+                        string.Format(
+                            "The UTF-16 size of the JSON-serialized payload must not exceed {0:N0} KB. The current payload size is {1:N0} KB.",
+                            maxSizeInKB,
+                            payloadSizeInKB));
+                }
+            }
```
Comment (tohling) [inline src/WebJobs.Extensions.DurableTask/MessagePayloadDataConverter.cs]: Why was this added back?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs [hunk 2]
```diff
@@ -59,20 +61,24 @@ internal class TaskActivityShim : TaskActivity
...
                     this.activityName,
                     this.activityVersion,
                     instanceId,
-                    result.Exception?.ToString() ?? string.Empty,
+                    exceptionToReport?.ToString() ?? string.Empty,
                     functionType: FunctionType.Activity,
                     isReplay: false);
 
-                if (result.Exception != null)
+                if (exceptionToReport != null)
```
Comment (tohling) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: >null [](start = 41, length = 4)

If for some reason the strip failed, should we log internally what the exception was?

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: I don't think we actually ever expect `exceptionToReport` to be null. I'm not sure what you mean when you say "for some reason the strip failed". Can you elaborate?

---
In reply to: [181521966](https://github.com/Azure/azure-functions-durable-extension/pull/250#discussion_r181521966) [](ancestors = 181521966)

Comment (tohling) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: I was referring to the method StripFunctionInvocationException. If we expect 

e as FunctionInvocationException to never return null, then it's OK.

---
In reply to: [181523045](https://github.com/Azure/azure-functions-durable-extension/pull/250#discussion_r181523045) [](ancestors = 181523045,181521966)

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: Right - I'm saying I don't think that method will ever return null. `e as FunctionInvocationException` might return null, but in that case we return `e` directly, so there shouldn't be any nulls. I can remove the `?` operator if that would make things less confusing.

---
In reply to: [181523423](https://github.com/Azure/azure-functions-durable-extension/pull/250#discussion_r181523423) [](ancestors = 181523423,181523045,181521966)

Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: Actually, I'll probably keep the null checks as they are. I don't have any guarantee that `result.Exception` will be non-null when `result.Succeeded` is false since that depends on the behavior of WebJobs. But to your original concern, I don't think there is ever a case where this is an exception and we fail to log it.

---
In reply to: [181523733](https://github.com/Azure/azure-functions-durable-extension/pull/250#discussion_r181523733) [](ancestors = 181523733,181523423,181523045,181521966)

File: src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs [hunk 3]
```diff
@@ -59,20 +61,24 @@ internal class TaskActivityShim : TaskActivity
...
                 {
-                    // Preserve the original exception context so that the durable task
-                    // framework can report useful failure information.
-                    ExceptionDispatchInfo.Capture(result.Exception).Throw();
+                    throw new TaskFailureException(
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs]: >throw new TaskFailureException( [](start = 20, length = 31)

We throw `TaskFailureException` here because that is the only exception type where the Durable Task Framework will actually try to deserialize the inner exception back in the orchestrator function.

File: test/TestHelpers.cs [hunk 4]
```diff
@@ -298,13 +298,13 @@ private static List<string> GetLogs_UnhandledActivityException(string messageId,
             {
                 $"{messageId}: Function '{orchestratorFunctionNames[0]} ({FunctionType.Orchestrator})', version '' scheduled. Reason: NewInstance. IsReplay: False.",
                 $"{messageId}: Function '{orchestratorFunctionNames[0]} ({FunctionType.Orchestrator})', version '' started. IsReplay: False. Input: \"Kah-BOOOOM!!!\"",
-                $"{messageId}: Function '{activityFunctionName} ({FunctionType.Activity})', version '' scheduled. Reason: Throw. IsReplay: False.",
+                $"{messageId}: Function '{activityFunctionName} ({FunctionType.Activity})', version '' scheduled. Reason: ThrowOrchestrator. IsReplay: False.",
                 $"{messageId}: Function '{orchestratorFunctionNames[0]} ({FunctionType.Orchestrator})', version '' awaited. IsReplay: False.",
                 $"{messageId}: Function '{activityFunctionName} ({FunctionType.Activity})', version '' started. IsReplay: False. Input: [\"Kah-BOOOOM!!!\"]",
-                $"{messageId}: Function '{activityFunctionName} ({FunctionType.Activity})', version '' failed with an error. Reason: Microsoft.Azure.WebJobs.Host.FunctionInvocationException",
```
Comment (tohling) [inline test/TestHelpers.cs]: >Microsoft.Azure.WebJobs.Host.FunctionInvocationException [](start = 133, length = 56)

The 'Reason: Microsoft.Azure.WebJobs.Host.FunctionInvocationException' entry in all these logs is also giving me a quick clue that the problem is a FunctionInvocationException, vs FunctionTimeoutException. In long and complex execution workflows, it may help quickly identify issues. We lose this visibility with this change. Do you think we really don't need to surface this information?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs [hunk 5]
```diff
@@ -81,27 +84,31 @@ public void SetFunctionInvocationCallback(Func<Task> callback)
...
                         this.context.Version,
                         this.context.InstanceId,
-                        e.ToString(),
+                        exceptionDetails,
                         FunctionType.Orchestrator,
                         this.context.IsReplaying));
                 }
 
-                throw;
+                throw new OrchestrationFailureException(
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs]: >OrchestrationFailureException [](start = 26, length = 29)

`OrchestrationFailureException` is the equivalent of `TaskFailureException` but for orchestrations and sub-orchestrations. We need to throw this exception specifically to ensure that the exception details are surfaced to the calling orchestrator function.

File: test/TestHelpers.cs (hunk 4 above)
Comment (cgillum) [reply test/TestHelpers.cs]: I've never heard of `FunctionTimeoutException`. If a function takes to long to execute, WebJobs will recycle the host. But even if we do have other exception types, I'm only ever stripping `FunctionInvocationException`, so you should still be able to figure out what's going on, right? In any case, customers have never needed to deal with these internal WebJobs exceptions, so I didn't want to surface them here.

---
In reply to: [181524809](https://github.com/Azure/azure-functions-durable-extension/pull/250#discussion_r181524809) [](ancestors = 181524809)
